metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/storage/LedgerStorage.scala

```diff
@@ -0,0 +1,42 @@
+package io.iohk.metronome.checkpointing.service.storage
+
+import io.iohk.metronome.checkpointing.models.Ledger
+import io.iohk.metronome.storage.{KVRingBuffer, KVCollection, KVStore}
+import scodec.Codec
+
+/** Storing the committed and executed checkpoint ledger.
+  *
+  * Strictly speaking the application only needs the committed state,
+  * since it has been signed by the federation and we know it's not
+  * going to be rolled back. Uncommitted state can be kept in memory.
+  *
+  * However we want to support other nodes catching up by:
+  * 1. requesting the latest Commit Q.C., then
+  * 2. requesting the block the Commit Q.C. points at, then
+  * 3. requesting the ledger state the header points at.
+  *
+  * We have to allow some time before we get rid of historical state,
+  * so that it doesn't disappear between step 2 and 3, resulting in
+  * nodes trying and trying to catch up but always missing the beat.
+  *
+  * Therefore we keep a collection of the last N ledgers in a ring buffer.
+  */
+class LedgerStorage[N](
+    ledgerColl: KVCollection[N, Ledger.Hash, Ledger],
+    ledgerMetaNamespace: N,
+    maxHistorySize: Int
```

Comment:
How do we choose this number? Could we not use the committed state as a cutoff point instead?

Reply:
I'm not sure, I guess we just have to find a value that is high enough so we don't see an errors logged. If we have a block every 5 seconds, then maybe save an hours' worth of state. Similarly to block pruning. I think I wanted to keep these two separate so that even if someone doesn't use block pruning, we still don't keep around state forever after it's been committed. 

If I understand the question, you're asking whether we could remove the state as soon as it's committed? Or at the next commit? Again I just want to avoid that if we get a new state every 1 second, then we get 404 every time we try to sync with someone. This fixed size seemed easier to implement than a timestamp based solution, but that's all I really wanted to achieve, have a few minute window to ask for history.

Reply:
Assuming we're storing only committed state there should always be a state to respond with to sync queries. Why would there be a 404? Why would we need hours of historical data?

Reply:
I mean if we only store the _last_ committed state and it changes quicker than how another node that realises that it's out of sync would fetch it before it's overwritten by the next commit. 

Say Alice and Bob are in sync and they keep producing new commits every 1-5 seconds, and the round timeout is 10 seconds. Charlie sees that it got no valid messages in a round when he times out, and starts syncing with the others. He asks everyone for their last commit quorum certificate, which can take some time if someone is not responding. Then he asks them for the state that corresponds to the last committed block as well -> if at that point everyone the state is no longer available because there was a new commit between Alice and Bob, he has to start again. 

So we don't need hours, just enough for that not to happen.

Reply:
It could be solved by other means like Guava caching; the ring buffer was just an idea I came up with to make sure we don't end up spinning our wheels with no control over it due to timing issues. Just want it to be a reasonable expectation that if Alice tells Charlie her latest block is X, then Charlie can follow up with the question of the state at that block is without running into data availability issues.

Reply:
Ok, I did not consider querying for QC and the associated state would be separate queries, but yeah, it does make sense.

Btw, I'm not questioning the ring buffer, I really like it :)

Reply:
Cool; at some point I should add separate tests for it, just made it generic because I wanted to see this working with the Robot example, so it's a bit more representative of what could be an approach that involves state sync.